openedx/core/lib/block_structure/transformer.py

```diff
@@ -125,4 +126,55 @@ def transform(self, usage_info, block_structure):
                 block structure, with already collected data for the
                 transformer, that is to be transformed in place.
         """
-        pass
+        raise NotImplementedError
+
+
+class OptimizedTransformer(BlockStructureTransformer):
+    """
+    Transformers may optionally choose to subclass this class instead of the
+    base, if their transform logic can be broken apart into a lambda for
+    optimization of combined tree traversals.
+
+    For performance reasons, developers should try to subclass this
+    class instead of the above base class, whenever possible - since
+    with this alternative, traversal of the entire block structure happens
+    only once for all transformers that implement this form of transform.
+    """
+
+    def transform(self, usage_info, block_structure):
+        """
+        This override of the abstract base method allows OptimizedTransformers
+        to define transform_block_filter instead of transform.
+        """
+        block_structure.filter_topological_traversal(self.transform_block_filter(usage_info, block_structure))
+
+    @abstractmethod
+    def transform_block_filter(self, usage_info, block_structure):
+        """
+        This is an alternative to the standard transform method.
+
+        Returns a filter function to be used to filter out any unwanted blocks
```

Comment:
It seems this commit makes it so this method can return a list of filters now, not just a single filter.